src/SharedStream.mo

```diff
@@ -0,0 +1,380 @@
+import Debug "mo:base/Debug";
+import Error "mo:base/Error";
+import Int "mo:base/Int";
+import Iter "mo:base/Iter";
+import Nat "mo:base/Nat";
+import Option "mo:base/Option";
+import Principal "mo:base/Principal";
+import R "mo:base/Result";
+import Time "mo:base/Time";
+import Timer "mo:base/Timer";
+
+import Vec "mo:vector";
+import QueueBuffer "QueueBuffer";
+
+module {
+
+  type StreamError = { #NotRegistered; #StreamClosed };
+
+  /// Usage:
+  ///
+  /// let receiver = StreamReceiver<Int>(
+  ///   123
+  ///   func (streamId : Nat, element: Int, index: Nat): () {
+  ///     ... do your logic with incoming item
+  ///   }
+  /// );
+  ///
+  /// Hook-up receive function in the actor class:
+  /// public shared func onStreamChunk(streamId : Nat, chunk: [Int], firstIndex: Nat) : async () {
+  ///   switch (streamId) case (123) { await receiver.onChunk(chunk, firstIndex); }; case (_) { Error.reject("Unknown stream"); }; };
+  /// };
+  public class StreamReceiver<T>(
+    streamId : Nat,
+    callback : (streamId : Nat, item : T, index : Nat) -> (),
+    startFromIndex : Nat,
+  ) {
+
+    var expectedNextIndex_ : Nat = startFromIndex;
+
+    var lastChunkTimestamp : Time.Time = Time.now();
+
+    public func isStreamClosed() : Bool = (Time.now() - lastChunkTimestamp) > 120_000_000_000; // 2 minutes
+
+    public func onChunk(chunk : [T], firstIndex : Nat) : R.Result<(), { #StreamClosed }> {
+      if (isStreamClosed()) {
+        return #err(#StreamClosed);
+      };
+      if (firstIndex != expectedNextIndex_) {
+        Debug.trap("Broken chunk index: " # Nat.toText(firstIndex) # "; expected: " # Nat.toText(expectedNextIndex_));
+      };
+      for (index in chunk.keys()) {
+        callback(streamId, chunk[index], firstIndex + index);
+      };
+      expectedNextIndex_ += chunk.size();
+      #ok();
+    };
+
+  };
+
+  /// Usage:
+  ///
+  /// let sender = StreamSender<Int>(
+  ///   123,
+  ///   10,
+  ///   10,
+  ///   func (item) = 1,
+  ///   anotherCanister.appendStream,
+  /// );
+  /// sender.next(1);
+  /// sender.next(2);
+  /// .....
+  /// sender.next(12);
+  /// await* sender.sendChunk(); // will send (123, [1..10], 0) to `anotherCanister`
+  /// await* sender.sendChunk(); // will send (123, [11..12], 10) to `anotherCanister`
+  /// await* sender.sendChunk(); // will do nothing, stream clean
+  public class StreamSender<T>(
+    streamId : Nat,
+    maxSize : ?Nat,
+    weightLimit : Nat,
+    weightFunc : (item : T) -> Nat,
+    callback : (streamId : Nat, items : [T], firstIndex : Nat) -> async R.Result<(), StreamError>,
```

Comment:
Make max concurrent chunks a constructor argument.

Reply:
We can simplify and generalize the arguments maxSize, weightLimit, weightFunc into one argument. All we need is a selection predicate which is a function T -> Bool. We could call this `selector : T -> Bool`. The user of this library can pass in a function that is part of an object that tracks number of elements (maxSize) or weight of elements, whatever the user wants.

Reply:
For the max concurrent chunks we also need a setter function so the user of the library can change it while the stream is active (not only in the constructor argument).

Reply:
Regarding generalizing the arguments maxSize, weightLimit, weightFunc, pros of having it included into SharedStream:
- cross-canister requests are always bound by max body size, so either "max amount" or "max size" (with having ability to calculate size) is required to all use-cases: current implementation allows to handle it easily: for max amount you can just set weight 1 for every item and set wightLimit to desired amount
- it's not so convenient in top level application: for regular use-case, app has to accumulate values to some variable, and this variable has to be zeroed before each chunk sending, so concurrency will break that completely. If define accumulator variable in each chunk sending, we will have to modify `selecter` function each time, which is also breaking concurrency

IMO current solution is flexible enough and makes things easier, what do you think?

Reply:
I see, ok. I also confused what maxSize was. Ok, then we leave it as it is. My comments are then:
- more descriptive name for maxSize (for example maxQueueSize, queueLimit, etc.)
- have multiple setter function that allow changing after the stream has been constructed at least the values of weightLimit and maxConcurrentChunks (because they may depend on network conditions, for example maxSize doesn't and I think we don't need to change it dynamically
- the timeout before sending empty batches

Reply:
done